src/Microsoft.DocAsCode.Build.Engine/Incrementals/DependencyGraph.cs

```diff
@@ -195,35 +196,42 @@ public static DependencyGraph Load(TextReader reader)
 
         private void RegisterDependencyTypeCore(DependencyType dt)
         {
-            DependencyType stored;
-            if (_types.TryGetValue(dt.Name, out stored))
+            lock (_typeSync)
             {
-                // to-do: add check for phase when new value overwrites old value
-                if (stored.IsTransitive != dt.IsTransitive)
+                DependencyType stored;
+                if (_types.TryGetValue(dt.Name, out stored))
                 {
-                    Logger.LogError($"Dependency type {JsonUtility.Serialize(dt)} isn't registered successfully because a different type with name {dt.Name} is already registered. Already registered one: {JsonUtility.Serialize(stored)}.");
-                    throw new InvalidDataException($"A different dependency type with name {dt.Name} is already registered");
-                }
-                if (stored.Phase == null)
-                {
-                    stored.Phase = dt.Phase;
+                    // to-do: add check for phase when new value overwrites old value
+                    if (stored.IsTransitive != dt.IsTransitive)
+                    {
+                        Logger.LogError($"Dependency type {JsonUtility.Serialize(dt)} isn't registered successfully because a different type with name {dt.Name} is already registered. Already registered one: {JsonUtility.Serialize(stored)}.");
+                        throw new InvalidDataException($"A different dependency type with name {dt.Name} is already registered");
+                    }
+                    if (stored.Phase != null)
+                    {
+                        Logger.LogVerbose($"Same dependency type with name {dt.Name} has already been registered, ignored.");
+                        return;
+                    }
                 }
-                Logger.LogVerbose($"Same dependency type with name {dt.Name} has already been registered, ignored.");
-                return;
+                var types = new Dictionary<string, DependencyType>(_types);
+                types[dt.Name] = dt;
+                _types = types.ToImmutableDictionary();
```

Comment:
>_types [](start = 16, length = 6)

can _types = _types.SetItem(dt.Name, dt)? It will return a new ImmutableDictionary?